pkg/operation/botanist/component/kubeapiserver/hvpa.go

```diff
@@ -0,0 +1,207 @@
+	"k8s.io/utils/pointer"
+)
+
+func (k *kubeAPIServer) emptyHVPA() *hvpav1alpha1.Hvpa {
+	return &hvpav1alpha1.Hvpa{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.DeploymentNameKubeAPIServer, Namespace: k.namespace}}
+}
+
+func (k *kubeAPIServer) reconcileHVPA(ctx context.Context, hvpa *hvpav1alpha1.Hvpa, deployment *appsv1.Deployment) error {
+	if k.values.Autoscaling.HVPAEnabled &&
+		k.values.Autoscaling.Replicas != nil &&
+		*k.values.Autoscaling.Replicas > 0 {
+
+		var (
+			hpaLabels           = map[string]string{v1beta1constants.LabelRole: v1beta1constants.LabelAPIServer + "-hpa"}
+			vpaLabels           = map[string]string{v1beta1constants.LabelRole: v1beta1constants.LabelAPIServer + "-vpa"}
+			updateModeAuto      = hvpav1alpha1.UpdateModeAuto
+			scaleDownUpdateMode = updateModeAuto
+			containerPolicyOff  = autoscalingv1beta2.ContainerScalingModeOff
+			hpaMetrics          = []autoscalingv2beta1.MetricSpec{
+				{
+					Type: autoscalingv2beta1.ResourceMetricSourceType,
+					Resource: &autoscalingv2beta1.ResourceMetricSource{
+						Name:                     corev1.ResourceCPU,
+						TargetAverageUtilization: pointer.Int32(hpaTargetAverageUtilizationCPU),
+					},
+				},
+			}
+			vpaContainerResourcePolicies = []autoscalingv1beta2.ContainerResourcePolicy{
+				{
+					ContainerName: containerNameKubeAPIServer,
+					MinAllowed: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("300m"),
+						corev1.ResourceMemory: resource.MustParse("400M"),
+					},
+					MaxAllowed: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("8"),
+						corev1.ResourceMemory: resource.MustParse("25G"),
+					},
+				},
+				{
+					ContainerName: containerNameVPNSeed,
+					Mode:          &containerPolicyOff,
+				},
+			}
+			weightBasedScalingIntervals = []hvpav1alpha1.WeightBasedScalingInterval{
+				{
+					VpaWeight:         hvpav1alpha1.VpaOnly,
+					StartReplicaCount: k.values.Autoscaling.MaxReplicas,
+					LastReplicaCount:  k.values.Autoscaling.MaxReplicas,
+				},
+			}
+		)
+
+		if k.values.Autoscaling.UseMemoryMetricForHvpaHPA {
+			hpaMetrics = append(hpaMetrics, autoscalingv2beta1.MetricSpec{
+				Type: autoscalingv2beta1.ResourceMetricSourceType,
+				Resource: &autoscalingv2beta1.ResourceMetricSource{
+					Name:                     corev1.ResourceMemory,
+					TargetAverageUtilization: pointer.Int32(hpaTargetAverageUtilizationMemory),
+				},
+			})
+		}
+
+		if k.values.Autoscaling.ScaleDownDisabledForHvpa {
+			scaleDownUpdateMode = hvpav1alpha1.UpdateModeOff
+		}
+
+		if k.values.SNI.PodMutatorEnabled {
+			vpaContainerResourcePolicies = append(vpaContainerResourcePolicies, autoscalingv1beta2.ContainerResourcePolicy{
+				ContainerName: containerNameAPIServerProxyPodMutator,
+				Mode:          &containerPolicyOff,
+			})
+		}
+
+		if k.values.Autoscaling.MaxReplicas > k.values.Autoscaling.MinReplicas {
```

Comment:
Can maxReplicas be nil here?

Reply:
`MaxReplicas` is of type `int32`, it's not a pointer?

Reply:
yep, got confused by github.